notify-debouncer-mini/src/lib.rs

```diff
@@ -160,88 +158,126 @@ pub struct DebouncedEvent {
 }
 
 impl DebouncedEvent {
+    #[inline(always)]
     fn new(path: PathBuf, kind: DebouncedEventKind) -> Self {
         Self { path, kind }
     }
 }
 
-type DebounceData = Arc<Mutex<DebounceDataInner>>;
+enum InnerEvent {
+    NotifyEvent(Result<Event, Error>),
+    Shutdown,
+}
 
-#[derive(Default)]
 struct DebounceDataInner {
+    /// Path -> Event data
     d: HashMap<PathBuf, EventData>,
+    /// timeout used to compare all events against, config
     timeout: Duration,
-    e: Vec<crate::Error>,
+    /// last debounce run
+    last_run: Instant,
+    /// oldest buffered event
+    oldest_update: Instant,
 }
 
 impl DebounceDataInner {
+    pub fn new(timeout: Duration) -> Self {
+        let time = Instant::now();
+        Self {
+            timeout: timeout,
+            oldest_update: time,
+            last_run: time,
+            d: Default::default(),
+        }
+    }
+
+    /// Returns a duration to wait for the next tick
+    #[inline]
+    pub fn next_tick(&self) -> Option<Duration> {
+        if self.has_tick_data() {
+            let next_run = self.oldest_update.checked_add(self.timeout)
+                .map_or(self.timeout, |v|v.saturating_duration_since(Instant::now()));
+            Some(next_run)
+        } else {
+            None
+        }
+    }
+
+    /// Returns true if there are buffered events
+    #[inline]
+    pub fn has_tick_data(&self) -> bool {
+        !self.d.is_empty()
+    }
+
     /// Retrieve a vec of debounced events, removing them if not continuous
+    ///
+    /// Updates the internal tracker for the next tick
     pub fn debounced_events(&mut self) -> Vec<DebouncedEvent> {
         let mut events_expired = Vec::with_capacity(self.d.len());
         let mut data_back = HashMap::with_capacity(self.d.len());
         // TODO: perfect fit for drain_filter https://github.com/rust-lang/rust/issues/59618
+        // reset oldest update
+        self.oldest_update = Instant::now();
```

Comment:
While this doesn't have any implication, with the logic when you don't `Option` for future deadline that looks error prone. Would suggest to simply set to `None` and rewrite the computation around storing the future deadline.

Reply:
Changed